### DISCUSSION TITLE
Intel faster than AMD, why?

### DISCUSSION THREAD

**Discussion by y1901y** [openai/whisper#2468] [Q&A]
Why is performance on an AMD processor half as good as on an Intel processor?
After all, AMD processor is more productive by all parameters.

Testing performance on Intel and AMD processors, bench parameters:
Main:
OS - Windows Server 2022
Hypervisor - VMware ESXi 7.0 U3

Stand01:
CPU - AMD EPYC 9374F, 4 vCPU
RAM - 32 Gb

Stand02:
CPU - Intel Xeon Gold 6354, 4 vCPU
RAM - 32 Gb

Results:
 Model   | AMD | Intel
medium | 41.46s | 24.92s
large-v2 | 96.58s | 50.47s
large-v3 | 87.21s | 44.89s
turbo     | 108.06s | 51.17s

```import torch
import whisper
import os
import time
from faster_whisper import WhisperModel, BatchedInferencePipeline

test_file = 'audio.mp3'
devices = [ 'cpu' ]#, 'cuda' ]
model_list = [ 'medium', 'large-v2', 'large-v3', 'turbo' ]
compute_types = [ 'int8_float32',  'float32' ] #, 'int8_float16', 'float16', 'bfloat16' ]

print('faster_whisper implementations')
for y in devices:
	for i in model_list:
		for x in compute_types:
			model = WhisperModel(i, device=y, compute_type=x)
			batched_model = BatchedInferencePipeline(model=model)
			
			duration_sum = 0
			start = time.time()
			segments, info = batched_model.transcribe(test_file, language='ru', task='transcribe', batch_size=16)
			end = time.time()
			duration_sum = duration_sum + end - start
			print("{} - quant {} for {} model with fp16 costs {:.2f}s".format(y, x, i, duration_sum))
			del model
#########################
print('whisper implementations')
device = torch.device('cpu')
model_list = [ 'medium', 'large-v2', 'large-v3', 'turbo' ]
fp16_bool = [True, False]
for i in model_list:
	for k in fp16_bool:
		model = whisper.load_model(name=i, device=device)
		duration_sum = 0
		audio = whisper.load_audio('audio.mp3')
		start = time.time()
		result = model.transcribe(audio, language='en', task='transcribe', fp16=k)
		end = time.time()
		duration_sum = duration_sum + end - start
		print("{} model with fp16 {} costs {:.2f}s".format(i, k, duration_sum))
		del model
```

**Comment by Advait251206** [comment]
Your results are surprising at first glance, but there are several reasons why an Intel CPU can outperform an AMD CPU for Whisper workloads despite AMD having stronger overall specifications.

## 1. Whisper Is Not a Generic CPU Benchmark

Whisper (and especially `faster-whisper` / CTranslate2) is heavily dependent on:

- Matrix multiplications (GEMM)
- Vector instructions (AVX2, AVX-512, VNNI)
- Memory bandwidth
- Cache behavior
- Thread scheduling

A CPU that wins in general benchmarks does not necessarily win on transformer inference.

---

## 2. Intel May Be Using AVX-512 / VNNI

The biggest thing to check is whether your software stack is using:

```text
AVX-512
AVX512-VNNI
Intel MKL
oneDNN
```

on the Intel host.

The Xeon Gold 6354 supports AVX-512 and VNNI acceleration.

Many transformer inference libraries gain huge speedups from these instructions.

By contrast, the EPYC 9374F does not implement Intel's AVX-512 execution model.

As a result:

```text
Intel:
optimized kernel
↓
AVX-512 / VNNI
↓
much faster inference
```

while AMD may fall back to:

```text
AVX2
```

or less optimized kernels.

---

## 3. CTranslate2 Optimizations Favor Intel

`faster-whisper` is built on:

```text
CTranslate2
```

which relies on:

- oneDNN
- Intel-optimized kernels
- CPU-specific dispatch

Historically, Intel CPUs have often received more optimization effort than AMD CPUs.

The gap has narrowed in recent years, but it can still be substantial depending on versions.

---

## 4. VMware Can Distort CPU Features

You're running:

```text
VMware ESXi 7.0 U3
```

with:

```text
4 vCPU
```

on both systems.

This introduces another variable:

### Verify what the VM actually sees

Inside each VM run:

```bash
lscpu
```

or on Windows:

```powershell
wmic cpu get Name
```

and inspect supported instruction sets.

It's possible that:

```text
Intel VM:
AVX512 exposed
```

while

```text
AMD VM:
different feature set
```

or vice versa.

VMware CPU masking can significantly affect performance.

---

## 5. Memory Topology Matters

Whisper inference frequently moves large tensors.

A VM with:

```text
4 vCPU
```

does not necessarily get:

```text
4 physical cores
```

with equal NUMA locality.

Check:

```text
NUMA node placement
CPU affinity
Memory affinity
```

on both hosts.

Poor NUMA placement can easily cost 20–50%.

---

## 6. Threading Configuration

You are using:

```python
batch_size=16
```

but not explicitly controlling thread counts.

Try benchmarking with:

```bash
OMP_NUM_THREADS=4
MKL_NUM_THREADS=4
```

or:

```python
import torch
torch.set_num_threads(4)
```

Sometimes AMD systems suffer from suboptimal default threading.

---

## 7. Turbo Model Slower Than Large-v3?

One thing that stands out:

```text
large-v3  -> 87 s
turbo     -> 108 s
```

Turbo should generally be faster.

This suggests your benchmark may be bottlenecked by:

- model loading
- CPU kernel selection
- quantization backend
- threading

rather than raw decoder complexity.

I would separate:

```python
model = WhisperModel(...)
```

from:

```python
transcribe(...)
```

and measure only inference.

---

## 8. Verify Which Backend Is Being Used

For CTranslate2:

```python
import ctranslate2
print(ctranslate2.get_supported_compute_types("cpu"))
```

Also check:

```bash
python -c "import torch; print(torch.__config__.show())"
```

You may discover:

```text
Intel:
MKL + oneDNN
```

vs

```text
AMD:
OpenBLAS
```

which can produce exactly the kind of 2× difference you're seeing.

---

## Things I'd Check First

### On both VMs

```bash
lscpu
```

Look for:

```text
avx2
avx512f
avx512vnni
```

### PyTorch build

```python
import torch
print(torch.__config__.show())
```

### CTranslate2 version

```python
import ctranslate2
print(ctranslate2.__version__)
```

### CPU utilization

Monitor:

```bash
htop
```

or Windows Task Manager.

If the AMD VM is not saturating all vCPUs, the issue may be scheduling rather than CPU capability.

---

## Most Likely Explanation

Given the nearly consistent **~2× advantage** across all models, the most probable causes are:

1. **Intel AVX-512/VNNI optimizations being used**
2. **oneDNN/MKL kernels favoring Intel**
3. **VMware exposing different CPU features**
4. **Thread scheduling or NUMA differences**

The gap is too large to be explained by raw CPU specifications alone and strongly suggests a software or instruction-set optimization difference rather than a Whisper model issue.
